### PR TITLE
Views

### PR DESCRIPTION
**Pull request by avneetmalhotra**

### CODE REVIEW

File: app/models/product.rb [hunk 1]
```diff
@@ -42,6 +45,8 @@ class Product < ApplicationRecord
   ## image_url validations
   validates :image_url, allow_blank: true, image_url: true
 
+  validates_with CategoryIdsValidator
```
Comment (nimish13) [inline app/models/product.rb]: @avneetmalhotra Think of a better way of doing this

File: db/migrate/20171129121009_create_addresses.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class CreateAddresses < ActiveRecord::Migration[5.1]
+  def change
+    create_table :addresses do |t|
```
Comment (nimish13) [inline db/migrate/20171129121009_create_addresses.rb]: @avneetmalhotra What about `timestamps` ?

File: app/models/address.rb [hunk 3]
```diff
@@ -0,0 +1,3 @@
+class Address < ApplicationRecord
```
Comment (nimish13) [inline app/models/address.rb]: @avneetmalhotra Please add basic validations

File: app/models/user.rb [hunk 4]
```diff
@@ -1,7 +1,9 @@
 class User < ApplicationRecord
   has_many :orders, dependent: :destroy
   has_many :line_items, through: :orders
-
+  has_one :address
```
Comment (nimish13) [inline app/models/user.rb]: @avneetmalhotra What will happen when we destroy a user

File: app/models/user.rb [hunk 5]
```diff
@@ -1,7 +1,9 @@
...
+  accepts_nested_attributes_for :address
```
Comment (nimish13) [inline app/models/user.rb]: do not save a address if all fields are empty

File: app/views/users/_form.html.erb [hunk 6]
```diff
@@ -19,6 +19,14 @@
     <%= form.text_field :name, id: :user_name %>
   </div>
 
+  <div class="field">
+    Address:
+    <br> 
+    <%= form.fields_for :addresses do |address_form| %>
```
Comment (nimish13) [inline app/views/users/_form.html.erb]: @avneetmalhotra - shouldn't it be `:address`

File: app/models/user.rb (hunk 5 above)
Comment (nimish13) [inline app/models/user.rb]: I should be able to destroy an address also using the same user form

File: app/controllers/products_controller.rb [hunk 7]
```diff
@@ -5,6 +5,12 @@ class ProductsController < ApplicationController
   # GET /products.json
   def index
     @products = Product.all
+    # render json: @products
+    # render formats: [:json]
+    respond_to do |format|
+      format.html
+      format.json { render json: Product.joins(:categories).pluck(:title, 'categories.name') }
```
Comment (nimish13) [inline app/controllers/products_controller.rb]: use {[], [], []} this syntax

File: app/models/product.rb [hunk 8]
```diff
@@ -5,6 +5,8 @@ class Product < ApplicationRecord
   has_many :carts, through: :line_items
   has_many :categorizations, dependent: :destroy
   has_many :categories, through: :categorizations
+  has_many :images
```
Comment (nimish13) [inline app/models/product.rb]: what about `dependent` option

File: db/migrate/20171130062258_create_images.rb [hunk 9]
```diff
@@ -2,8 +2,6 @@ class CreateImages < ActiveRecord::Migration[5.1]
   def change
     create_table :images do |t|
```
Comment (nimish13) [inline db/migrate/20171130062258_create_images.rb]: timestamps?
also store file_path in db

File: app/models/user.rb [hunk 10]
```diff
@@ -1,6 +1,9 @@
 class User < ApplicationRecord
-  has_many :orders
-
+  has_many :orders, dependent: :destroy
+  has_many :line_items, through: :orders
+  has_one :address, dependent: :destroy
+  accepts_nested_attributes_for :address
```
Comment (nimish13) [inline app/models/user.rb]: use blank and destroy options

File: app/views/products/_product_details.html.erb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<td>
+  <%= image_tag(product.image_url, class: 'list_image') if product.image_url%>
+</td>
+
+<td>
+  <%= image_tag(product.images.first.file_path, class: 'list_image', alt: "#{@first_image_name}") if product.images.first %>
```
Comment (nimish13) [inline app/views/products/_product_details.html.erb]: please use product.images.present?

File: lib/tasks/port_legacy_products.rake [hunk 12]
```diff
@@ -0,0 +1,4 @@
+desc "Assign the first category to all products, which are already created"
+task :port_legacy_product => :environment do
+  Product.all.each { |product| product.categories << Category.first }
```
Comment (nimish13) [inline lib/tasks/port_legacy_products.rake]: run this for only those products which don't have any category

File: app/controllers/products_controller.rb [hunk 13]
```diff
@@ -5,6 +5,10 @@ class ProductsController < ApplicationController
   # GET /products.json
   def index
     @products = Product.all
```
Comment (nimish13) [inline app/controllers/products_controller.rb]: avoid n+1 queries

**Comment by avneetmalhotra** [comment]
closed in reference to #11

**Pull request closed without merging**